### ISSUE TITLE
ipa-replica-install fails on Importing RA key

### ISSUE THREAD

**Issue by konsumate** [freeipa/freeipa-container#497]
Command:
`docker run --name freeipa-server --rm -ti -h auth2.localdomain.local --sysctl net.ipv6.conf.all.disable_ipv6=0 -p 0.0.0.0:88:88 -p 0.0.0.0:464:464 -p 0.0.0.0:636:636 -p 0.0.0.0:123:123 -p 0.0.0.0:7389:7389 -p 0.0.0.0:389:389 --sysctl net.ipv6.conf.lo.disable_ipv6=0 -v /sys/fs/cgroup:/sys/fs/cgroup:ro  -v /srv/freeipa/ipa-data:/data:Z freeipa-fed36 ipa-replica-install --domain=localdomain.local --realm=LOCALDOMAIN.LOCAL --server=auth1.localdomain.local -P ... -W ...`

Is able to get quite far in the replication setup, but fails at a later step while trying to import ra-agent keys:

```
Configuring certificate server (pki-tomcatd)
  [1/2]: configure certmonger for renewals
  [2/2]: Importing RA key
[**    ] Job ipa-server-configure-first.service/start running (2min 8s / no limit)
Error storing key "keys/ra/ipaCert": CalledProcessError(Command ['/usr/libexec/ipa/custodia/ipa-custodia-ra-agent', '--import', '-'] returned non-zero exit status 1: 'Traceback (most recent call last):\n  File "/usr/libexec/ipa/custodia/ipa-custodia-ra-agent", line 8, in <module>\n    main(ra_agent_parser())\n  File "/usr/lib/python3.10/site-packages/ipaserver/secrets/handlers/pemfile.py", line 117, in main\n    common.main(parser, export_key, import_key)\n  File "/usr/lib/python3.10/site-packages/ipaserver/secrets/handlers/common.py", line 73, in main\n    func(args, tmpdir, **kwargs)\n  File "/usr/lib/python3.10/site-packages/ipaserver/secrets/handlers/pemfile.py", line 72, in import_key\n    ipautil.run(cmd, umask=0o027)\n  File "/usr/lib/python3.10/site-packages/ipapython/ipautil.py", line 599, in run\n    raise CalledProcessError(\nipapython.ipautil.CalledProcessError: CalledProcessError(Command [\'/usr/bin/openssl\', \'pkcs12\', \'-in\', \'/tmp/tmpmx8ori9z/import.p12\', \'-clcerts\', \'-nokeys\', \'-out\', \'/var/lib/ipa/ra-agent.pem\', \'-password\', \'file:/tmp/tmpmx8ori9z/passwd\'] returned non-zero exit status 1: \'Error outputting keys and certificates\\n002ECDE28A7F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:349:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()\\n\')\n')
  [error] FileNotFoundError: [Errno 2] No such file or directory: '/var/lib/ipa/ra-agent.key'
```
Does /var/lib/ipa/ra-agent.key or /var/lib/ipa/ra-agent.pem need to be part of some volume-data-list or volume-tmp-list ?

Touching both files before the installation reaches that step will not fail it - but both will be ofc empty

**Comment by adelton** [comment]
The whole `/var/lib/ipa/` is in `volume-data-list`, so that whole directory lives on the data volume:
```
$ docker exec freeipa-replica ls -dla /var/lib/ipa
lrwxrwxrwx. 1 root root 17 Nov 25 04:37 /var/lib/ipa -> /data/var/lib/ipa
```

You might want to try things with
```
readonly=--read-only tests/run-master-and-replica.sh freeipa/freeipa-server:fedora-36
```
-- it passes for me:
```
Configuring certificate server (pki-tomcatd)
  [1/2]: configure certmonger for renewals
  [2/2]: Importing RA key
Done configuring certificate server (pki-tomcatd).
```
```
$ docker exec freeipa-replica ls -la /var/lib/ipa/
total 8
drwxr-xr-x. 10 root root    240 Nov 27 09:04 .
drwxr-xr-x. 13 root root    260 Nov 25 04:37 ..
drwx------.  2 root root     40 Aug 24 13:20 backup
drwxr-xr-x.  2 root root     60 Nov 27 09:04 certs
drwx------.  2 root root     60 Nov 27 09:04 gssproxy
drwx------.  2 root root     60 Nov 27 09:04 passwds
drwxr-xr-x.  3 root root     60 Nov 27 09:08 pki-ca
drwx------.  2 root root     60 Nov 27 09:04 private
-r--r-----.  1 root ipaapi 1828 Nov 27 09:04 ra-agent.key
-r--r-----.  1 root ipaapi 1549 Nov 27 09:04 ra-agent.pem
drwx--x--x.  2 root root    240 Nov 27 09:04 sysrestore
drwx------.  2 root root     60 Nov 27 09:03 sysupgrade
```

I would suggest to focus on the full traceback that you got. After unpacking and unescaping that single long line, it says:
```
Error storing key "keys/ra/ipaCert": CalledProcessError(
        Command [
                '/usr/libexec/ipa/custodia/ipa-custodia-ra-agent', '--import', '-'
        ] returned non-zero exit status 1:
Traceback (most recent call last):
  File "/usr/libexec/ipa/custodia/ipa-custodia-ra-agent", line 8, in <module>
    main(ra_agent_parser())
  File "/usr/lib/python3.10/site-packages/ipaserver/secrets/handlers/pemfile.py", line 117, in main
    common.main(parser, export_key, import_key)
  File "/usr/lib/python3.10/site-packages/ipaserver/secrets/handlers/common.py", line 73, in main
    func(args, tmpdir, **kwargs)
  File "/usr/lib/python3.10/site-packages/ipaserver/secrets/handlers/pemfile.py", line 72, in import_key
    ipautil.run(cmd, umask=0o027)
  File "/usr/lib/python3.10/site-packages/ipapython/ipautil.py", line 599, in run
    raise CalledProcessError(
ipapython.ipautil.CalledProcessError: CalledProcessError(Command [
        '/usr/bin/openssl', 'pkcs12', '-in', '/tmp/tmpmx8ori9z/import.p12', '-clcerts', '-nokeys',
          '-out', '/var/lib/ipa/ra-agent.pem', '-password', 'file:/tmp/tmpmx8ori9z/passwd'] returned non-zero exit status 1:
Error outputting keys and certificates
002ECDE28A7F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:349:Global default library context, Algorithm (RC2-40-CBC : 0),
 Properties ()
')
')
```

So the error actually comes from the `openssl pkcs12` command.

Is it possible that your FreeIPA master is some very old version with keys using encryption that the latest OSes and OpenSSL versions no longer support (without extra options like `-legacy`)?

Is it possible that you would get the same error if you run your `ipa-replica-install` on some Fedora 36 host, completely outside of containers?

In other words, are you sure this issue is container specific and not related to the compatibility of your FreeIPA master and the replica that you try to setup?

**Comment by konsumate** [comment]
Ohey,

The migration was from a plain centos 8.5 with ipa 4.9.6 into a 4.9.10 containerized fed63 setup.
It's unclear wether openssl can not save or parse stuff due to container restrictions or due to different formats.
Indeed this could be very well some OS incompatibility.

Workaround was:
- Sneak in the previous ra-agent.key and ra-agent.pem from the origin into the new replica during initial replica install (the old one was decomissed afterwards anyway)

**Comment by adelton** [comment]
@abbra Is this RA key incompatibility something expected between CentOS 8 and Fedora 36?

**Comment by abbra** [comment]
This is known issue fixed in FreeIPA 4.9.9 or later. CentOS 8 Stream already contains a fix since RHEL 8.6 release (May 2022). There were two issues:
 - IPA side, fixed in FreeIPA 4.9.9, https://pagure.io/freeipa/issue/9101
 - PKI side, fixed in PKI 11.0, https://pagure.io/freeipa/issue/8969

CentOS 8.5 is not supported anymore, you need to switch to at least CentOS 8 Stream or a similar version and then do the upgrade.

**Comment by adelton** [comment]
Thanks for the explanation.

@konsumate Can we consider this issue resolved?

**Comment by konsumate** [comment]
Yes - this issue can be closed, workaround included and issue seems to be known and fixed already.
